docs/zenith/developer/go/457202-test-build-publish.md

```diff
@@ -0,0 +1,303 @@
+
+```shell
+dagger shell package --entrypoint /bin/sh
+```
+
+This command drops you into an interactive shell, allowing you to directly execute commands in the container returned by the `package` function, as in the example below:
+
+```shell
+/ # cd /usr/share/nginx/html
+/usr/share/nginx/html # ls
+50x.html             favicon.ico          logo192.png          manifest.json        static
+asset-manifest.json  index.html           logo512.png          robots.txt
+/usr/share/nginx/html #
+```
+
+### dagger up
+
+The `dagger up` command allows `Service` and `Container` types returned by a function to be executed and have any exposed ports forwarded to the host machine. This has many potential use cases, such as manually testing web servers or databases directly from the host browser or host system.
+
+In order for this to work, the service/container returned by the function must have the `Container.withExposedPort` field defining one or more exposed ports. So, add a new `PackageService()` function as shown below:
+
+```go
+func (m *Mymod) PackageService(nodeVersion Optional[string]) *Service {
+  return m.Package().
+    WithExposedPort(8080).
+    AsService()
+}
+```
+
+Then, use the `dagger up` command to build the application and serve it with NGINX:
+
+```shell
+dagger up --native package-service
+```
+
+You should now be able to access the application by browsing to `http://localhost:8080` on the host (replace `localhost` with your Docker host's network name if accessing it remotely).
+
+For reference, here is the final module code:
+
+```go file=./snippets/test-build-publish/step6/main.go
+```
+
+## Conclusion
+
+This guide walked you through the process of creating a Dagger module from scratch to test, build and publish a Node.js application image. It explained how to create a module, add functions to it, and work with containers and services as function inputs and outputs. It also demonstrated how to use modules from the Daggerverse to speed up your module development, by reusing functionality developed by the Dagger community.
+
+Continue your journey into Dagger programming with the following resources:
+
+- The [Daggerverse](https://daggerverse.dev), an online catalog of Dagger modules for you to use and learn from
+- Guide on [advanced module programming](./191108-advanced-programming.md)
+- Guide on [publishing modules to the Daggerverse](../821742-publishing-modules.md)
+- Reference documentation for the [Go](https://pkg.go.dev/dagger.io/dagger) SDK
+
+## Appendix A: Create an example application
+
+This tutorial assumes that you have a Node.js Web application. If not, follow the steps below to create an example React application.
+
+1. Create a directory for the application:
+
+  ```shell
+  mkdir myapp
+  cd myapp
+  ```
+
+1. Create a skeleton application:
+
+  ```shell
+  npx create-react-app .
+  ```
+
+1. Make a minor modification to the application's index page:
+
+  ```shell
+  sed -i -e 's/Express/Dagger/g' routes/index.js
```

Comment:
this failed for me, there isn't a `routes/index.js` file

Reply:
Modified